File: stack/imp_list.py
class N:
    def __init__(self, d):
        self.d = d
        self.n = None

class S:
    def __init__(self):
        self.t = None

    def push(self, d):
        node = N(d)
        node.n = self.t
        self.t = node

    def pop(self):
        if not self.t:
            return None
        x = self.t.d
        self.t = self.t.n
        return x

    def peek(self):
        return self.t.d if self.t else None

    def empty(self):
        return self.t is None

    def disp(self):
        c = self.t
        while c:
            print(c.d, end=' ')
            c = c.n
        print()
stk = S()
stk.push(10)
stk.push(20)
stk.push(30)
stk.disp()        
print(stk.pop())  
print(stk.peek()) 
stk.disp()        

File: stack/imp_list.c
#include <stdio.h>
#include <stdlib.h>

typedef struct N {
    int d;
    struct N* n;
} N;

typedef struct S {
    N* t;
} S;

// Push
void push(S* s, int x) {
    N* p = (N*)malloc(sizeof(N));
    p->d = x;
    p->n = s->t;
    s->t = p;
}

// Pop
int pop(S* s) {
    if (!s->t) {
        printf("Stack Underflow\n");
        return -1;
    }
    int x = s->t->d;
    N* temp = s->t;
    s->t = s->t->n;
    free(temp);
    return x;
}

// Peek
int peek(S* s) {
    return s->t ? s->t->d : -1;
}

// Check empty
int empty(S* s) {
    return s->t == NULL;
}

// Display
void disp(S* s) {
    N* c = s->t;
    while (c) {
        printf("%d ", c->d);
        c = c->n;
    }
    printf("\n");
}

int main() {
    S stk = {NULL};
    push(&stk, 10);
    push(&stk, 20);
    push(&stk, 30);

    disp(&stk);          
    printf("%d\n", pop(&stk));   
    printf("%d\n", peek(&stk));  
    disp(&stk);          

    return 0;
}
